### ISSUE TITLE
Add option to return non-reranked results if an error occurs during reranking

### ISSUE THREAD

**Issue by pmpailis** [elastic/elasticsearch#116796]
Currently, when using the `text_similarity_reranker` retriever, if something goes wrong during the rerank process, whether this is a direct error from the reranking service or issues due to changes in response etc, we propagate this exception all the way up and return a 4xx response to the user. 

We could introduce a new parameter as part of the `text_similarity_reranker` retriever (defaults to `false`) which would allow "lenient reranking", i.e.  give the option to return the results non reranked if something goes wrong during the rerank process. 

We need to also take into account logging (as we don't want to pollute our logs with thousands of identical error logs) and how to actually notify the user that the results are not actually reranked.

**Comment by Danouchka** [comment]
Yes it could be also that your quota on CSP provider may be exhausted. so it's all the more important to return results even if not reranked...because we cannot imagine a RAG sending back errors to customers all the time

**Comment by kunal5042** [comment]
Hi, 
I would like to contribute to this, can it be assigned to me?

**Comment by Danouchka** [comment]
Hi @pmpailis any news regarding this ER please ?

**Comment by thecoop** [comment]
This work can be split into two independent parts:

1. Making `RankFeaturePhase.onPhaseDone` gracefully handle reranker failures. Rather than specific to the ML reranker, it should be a general option on reranking itself to specify the reranker as optional (although we only have one implementation, there should be nothing stopping us from using the same optionality for any later ones). To start with, this can just log the failure and drop the exception.
2. Add information on `SearchResponse` to report on failures. There's a few different ways we could do this:
    1. Add a REST header to the response indicating there was a failure (not sure how to do this)
    2. Expand `shardFailures` field to be able to store failures not specific to a shard. This lets the response return results, but also include a failure indicating reranker failure
    3. Add a new `subsidiaryFailures` field to store failure information not specific to a shard. To start with, this could only contain reranker failures, but could be used for any subsidiary failure that does not fail results completely.

**Comment by Danouchka** [comment]
Hello is this available now please ? In which release ?

**Comment by Danouchka** [comment]
If  there's no results at the query before going to the text similarity reranker itself then the reranker fails  as well 
GET search_chamonix_v2/_search
{
  "size": 10,
  "_source": {
    "excludes": [
      "images.image_embeddings",
      "description.semantic_e5",
      "description.semantic_minilm",
      "description.semantic_openai3large"
    ]
  },
  "script_fields": {
    "nb_images": {
      "script": {
        "source": "int i=0;for(item in params._source.images){i++} return i;"
      }
    }
  },
  "fields": [
    "partial_address",
    "nb_images"
  ],
  "retriever": {
    "text_similarity_reranker": {
      "retriever": {
        "rrf": {
          "query": "chalet avec une cuisine italienne, une suite parentale, vue sur l'aiguille du midi",
          "fields": [
            "description",
            "description.semantic_openai3large"
          ],
          "filter": {
            "bool": {
              "must": [
                {
                  "bool": {
                    "should": [
                      {
                        "match": {
                          "partial_address": {
                            "query": "Les Bossons",
                            "fuzziness": "AUTO"
                          }
                        }
                      }
                    ],
                    "minimum_should_match": 1
                  }
                }
              ],
              "should": [],
              "filter": [
                {
                  "range": {
                    "nb_chambres": {
                      "gte": 3
                    }
                  }
                }
              ]
            }
          }
        }
      },
      "field": "description",
      "inference_id": "cohere_rerank_prod",
      "inference_text": "chalet avec une cuisine italienne, une suite parentale, vue sur l'aiguille du midi"
    }
  }
}


Will lead to this error in my case because 0 results to rank:
{
  "error": {
    "root_cause": [
      {
        "type": "action_request_validation_exception",
        "reason": "Validation Failed: 1: [rank] requires [size] greater than [0];"
      }
    ],
    "type": "action_request_validation_exception",
    "reason": "Validation Failed: 1: [rank] requires [size] greater than [0];"
  },
  "status": 400
}

**Comment by Danouchka** [comment]
cc @dustincoates

**Comment by benwtrent** [comment]
it is not completed yet.

**Comment by Danouchka** [comment]
Hi @benwtrent  
When will it be released please ?

**Comment by thecoop** [comment]
There is currently no expected date this will be completed by

**Comment by Danouchka** [comment]
Hello @thecoop , Is there a possibility to implement it soon please ? because it seems you have the code in some merge if I understood well. Is there an issue with it ?

**Comment by thecoop** [comment]
I was working on this in some extra time I had; it is not part of any particular roadmap. I could not get it completed in time before I had to work on other areas, so it's waiting until I have some spare time again.
